server/jetstream_api.go

```diff
@@ -1047,6 +1048,30 @@ func (s *Server) getRequestInfo(c *client, raw []byte) (pci *ClientInfo, acc *Ac
 	return &ci, acc, hdr, msg, nil
 }
 
+func (s *Server) unmarshalRequest(pci *ClientInfo, acc *Account, subject string, msg []byte, v interface{}) error {
+	decoder := json.NewDecoder(bytes.NewReader(msg))
+	decoder.DisallowUnknownFields()
+
+	for {
+		err := decoder.Decode(v)
+
+		if err == io.EOF {
+			return nil
+		}
+
+		if err != nil {
+			s.Errorf("Invalid JetStream request '%s > %s': %s", acc, subject, err)
+
+			var config = s.JetStreamConfig()
+			if config.Strict {
+				return err
+			}
+
+			return json.Unmarshal(msg, v)
```

Comment:
Not convinced by this block, in that if strict mode is disabled, we can in effect unmarshal twice. (Unmarshalling JSON is actually quite expensive on CPU so need to be careful on hot paths.)

Why this vs just having one `decoder` and setting `DisallowUnknownFields()` based on the `Strict` setting?

Reply:
A misbehaving client is going to a) spam log lines and b) unmarshal twice for each request. I don't think that's wise, needs further discussion.

Reply:
Derek suggested that for 2.11 while we are in soft mode we should still report the problem so client authors can discover and remediate issues so he suggested this approach.  It's nice a nice assist before we enable this by default to reject requests but I agree with you it will spam logs until things are fixed.

Reply:
I think client authors could enable this as an opt-in too, in order to avoid spamming there is `rateLimitFormatWarnf` which would log less

Reply:
rateLimitFormatWarnf is a good idea though for sure

Reply:
I think that any opt-in for the client misses the point that Derek had - it's about detecting problems with non-tier-1 and older clients, and those will not have such opt-in implemented.